src/raft_server.c

```
@@ -1168,6 +1180,11 @@ int raft_append_entry(raft_server_t* me_, raft_entry_t* ety)
 {
     raft_server_private_t* me = (raft_server_private_t*)me_;
 
+    if (me->current_term < ety->term) {
+        // figure out error case
```

This cannot happen according to raft. For these kind of situations, RAFT_ERR_SHUTDOWN is used in the codebase :(e.g in `raft_recv_appendentries()`)

```c
            if (ae->prev_log_idx <= raft_get_commit_idx(me_))
            {
                /* Should never happen; something is seriously wrong! */
                raft_log_node(me_, ae->leader_id,
                            "AE prev conflicts with committed entry");
                e = RAFT_ERR_SHUTDOWN;
                raft_entry_release(ety);
                goto out;
            }
```

You can print a log and return RAFT_ERR_SHUTDOWN if you want.

---

so jut to be clear, you're ok with the change, the comment is just what the error return should be?  I agree it shouldn't happen.  The point is we have to include the test to make sure we write correct tests.

---

> you're ok with the change, the comment is just what the error return should be?

Yes, exactly.

---

so this turns out to be slightly more annoying.  if we return RAFT_ERR_SHUTDOWN, the caller should shutdown, but we end up eating the error and not returning it to client in most places.   so have to plumb that through everywhere.

---

I've checked `raft_server.c`, looks like we return raft_append_entry() errors to users. Are you talking about tests?

---

raft_append_entry() is not called by the user.  raft_recv_entry (which calls raft_append_entry on master), raft_recv_appendentries() (which calls raft_append_entry on follower) and raft_recv_requestvote_response() which calls raft_append_entry (noop) in context of raft_become_leader().  they don't plumb it back.

perhaps the best answer is that this should be an assert not return error?  i.e. we just die if you append an entry with a term > existing term?

would simplify tests, as can remove all the checks I added for raft_append_entry error() == 0

---

> raft_append_entry() is not called by the user. raft_recv_entry (which calls raft_append_entry on master), raft_recv_appendentries() (which calls raft_append_entry on follower) and raft_recv_requestvote_response() which calls raft_append_entry (noop) in context of raft_become_leader(). they don't plumb it back.

But all these three functions propagate the error through the API. Are we talking about the same thing? For example, this is `raft_become_leader()` propagates `raft_append_entry()`'s return code : 

https://github.com/RedisLabs/raft/blob/23f3b3bd3a1585ad71929fd17f10d7ba61d15b73/src/raft_server.c#L393



Anyway, I'm okay with assert as well. I agree with you, it makes more sense.

---

you're right.  I wasn't clear and confused myself in my description.  The plumbing that would be needed would be in redisraft.  i.e. they would have to check in all the situations for the error code.  they currently don't check in most places for RAFT_ERR_SHUTDOWN

but I'll change to an assert.